designs/mfa_backup_code.rst

```diff
@@ -17,65 +17,35 @@ To implement this, we can add Backup code as a new AuthCredential, so that it ca
 
 To validate Backup code, we simply check if the input matches the Backup code saved in database, which should be hashed with salt (similar to how password is validated).
 
-A user can only have one Backup code, if user wants a new Backup code:
-
-- if the existing Backup code is still valid, cannot create new Backup code
-- if no existing/valid Backup code, create/replace with a new Backup code
-
 The login session should only last for 5 mins with a Claim that allows users to change login credentials.
 
 Code Generation
 ------------------------
-The Backup code could be generated by TOTP/HOTP, or a cryptographically secure pseudorandom number generator. 
-Attackers should not be able to guess it even if they know our source code.
-
-Implementation 1
-------------------------
 
-Based on the Microsoft temporary access pass:
+The Backup code could be generated by TOTP/HOTP, or a cryptographically secure pseudorandom number generator. Attackers should not be able to guess it even if they know our source code.
 
-Add a new class/object: Backup code policy
+(See kanidmd/src/lib/utils.rs readable_password_from_random)
 
-- Each Kanidm has a Backup code policy
-- Define which users/groups can use the Backup code
-   - User can only login if they are included in the Backup code policy
-- Define parameters like max/min lifetime, one-time use and length
-- Only ``idm_hp_account_manage_priv`` can modify it?
+Implementation
+------------------------
 
 Add a new class/object: Backup code
 
-- Issue and save the Backup code in the database, it must adhere to Backup code policy
-- ``idm_account_manage_priv`` can create/read/delete it for normal users, but not themselves?
-- ``idm_hp_account_manage_priv`` can create/read/delete it for ``idm_account_manage_priv`` users and normal users?
+A user can generate one batch of Backup codes at a time, which will be saved in the database
+   - Each code in the batch can be used only once
+   - If the user generates a new batch of Backup codes, the last batch of codes become invalid.
 
-Workflow
+To invalidate the Backup code after usage, we should send an async message to the async task queue. Since the auth process is currently in a single transaction lock that is NOT writeable, using an async action allows the auth process to proceed in parallel and prevent opening nested transactions. **Note** that there is a small window between "use of the backup code" and the async action being processed to actually cause the invalidation, but this window is short enough that it's an acceptable compromise. 
```

Comment:
We also need to comment on how we need rate limiting/bruteforce protection against these codes similar to what exists for passwords.